protocols/mdns: Optimise `InterfaceState::poll` for low latency

thomaseizinger opened the pull request:
# Description

<!-- Please write a summary of your changes and why you made them.-->

This is now in accordance with our coding guidelines. Inspired by @gallegogt's findings.

## Links to any relevant issues

<!-- Reference any related issues.-->
- https://github.com/libp2p/rust-libp2p/discussions/2916

## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [x] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Rewrite `InterfaceState::poll` to have a single loop

This ensures we do as much work as possible before we need to get
polled again.

thomaseizinger committed:
Change priorities in poll loop

thomaseizinger committed:
Ensure to empty buffer even in case of error

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Bump version in manifest

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+
+- Fix a bug that could cause a delay of ~10s until peers would get discovered when using the tokio runtime. See [PR XXXX].
+
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
```suggestion
- Fix a bug that could cause a delay of ~10s until peers would get discovered when using the tokio runtime. See [PR 2939].

[PR 2939]: https://github.com/libp2p/rust-libp2p/pull/2939
```

thomaseizinger committed:
Update protocols/mdns/CHANGELOG.md

mxinden commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -200,51 +200,58 @@ where
+                self.send_buffer.push_back(build_query());
+            }
+
+            // 2nd priority: Keep local buffers small: Empty buffers ASAP.
```
Shouldn't we first try to send existing work (`self.send_buffer.pop_front`) and then add more work (`self.send_buffer.push_back`)? In case we do a `push_back`, we can do a `continue` to follow with a `pop_front`.

thomaseizinger replied:
I am not sure if it makes a big difference. The timer is our only local work, i.e. nothing else will push to the buffer. When we say "prefer local work over remote work", does that not imply that whatever generates the local work has the highest priority, followed by trying to send it to the remote?

mxinden commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -200,51 +200,58 @@ where
-        if Pin::new(&mut self.timeout).poll_next(cx).is_ready() {
-            log::trace!("sending query on iface {}", self.addr);
-            self.send_buffer.push_back(build_query());
+            // 4th priority: Emit event only if nothing else can make progress.
```
Would it not make more sense to do (4) before (3)? I.e. return discovered addresses before checking whether there are other to discover? That would keep the buffer `self.discovered` small.

thomaseizinger replied:
Yeah that would make sense I think!

mxinden reviewed (commented):
Thanks for looking into this.

I don't yet understand what the issue was. Was `self.timeout` not polled early on and thus no waker registered?

mxinden commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -200,51 +200,58 @@ where
         cx: &mut Context,
         params: &impl PollParameters,
     ) -> Option<(PeerId, Multiaddr, Instant)> {
```
Not directly related. Why do we return an `Option` and not a `Poll`?

thomaseizinger replied:
I don't know why.

thomaseizinger replied:
Fixed it!

thomaseizinger committed:
Move helper function to bottom

thomaseizinger committed:
Optimise for small, local buffers

thomaseizinger committed:
Return `Poll` from `poll` function

thomaseizinger committed:
Improve local variable name

thomaseizinger committed:
Use `VecDeque::extend` over manual iterator

thomaseizinger committed:
Extract helper for extracting responses

thomaseizinger committed:
Inline `inject_mdns_packet`

It is easier to understand this function if all buffer manipulations
happen in here instead of in helper functions.

thomaseizinger committed:
Flatten match statement

thomaseizinger committed:
Further flatten match statement

thomaseizinger committed:
MOAR flattening!

thomaseizinger committed:
Move `extract_discovered` to `MdnsResponse`

This results in better encapsulation.

thomaseizinger committed:
Extract time dependency

This allows for better unit testing. Not that we have tests for this
at the moment but in general :blush:

thomaseizinger committed:
Use `?` in `MdnsPacket::new` to propagate error

thomaseizinger committed:
Further flatten `MdnsPacket::new_from_bytes`

thomaseizinger committed:
Prefer early exit over if else

thomaseizinger committed:
Inline variables

thomaseizinger committed:
Improve logs

thomaseizinger committed:
Fix clippy

mxinden reviewed (approved):
Thank you @thomaseizinger.

thomaseizinger committed:
Merge branch 'master' into 2916-mdns-refactor-poll

thomaseizinger committed:
Put changelog items in correct order

thomaseizinger committed:
Fix bad merge of mDNS version

thomaseizinger committed:
Merge branch 'master' into 2916-mdns-refactor-poll

thomaseizinger committed:
Merge branch 'master' into 2916-mdns-refactor-poll
